File: scripts/secrets.py
#!/usr/bin/env python3
"""
Cortex secrets vault — v2 per-file format with v1 legacy read support.

Write: always per-file (cortex.secrets/<n>.enc)
Read:  per-file first, falls back to legacy blob (cortex.secrets.enc)
List:  union of both — per-file takes precedence on duplicate keys

DESIGN NOTE (2026-04-17):
  The original vault used a single encrypted blob (cortex.secrets.enc).
  This was a design mistake — a single blob cannot be merged by git, causing
  sync conflicts when two agents write simultaneously.

  v2 uses one encrypted file per secret (cortex.secrets/<n>.enc).
  Git merges per-file independently. Conflicts only occur if two agents write
  the SAME key at the same time — surfaces as a standard git conflict.

  The legacy blob reader below exists ONLY for backwards compatibility with
  repos that have not yet migrated. It is deprecated and will be removed in
  a future version. Run `python scripts/secrets.py migrate` to clean up.

Requires: pip install cryptography
"""

import os
import sys
import json
import base64
import getpass
import argparse
import datetime

try:
    from cryptography.hazmat.primitives.ciphers.aead import AESGCM
    from cryptography.hazmat.primitives.kdf.pbkdf2 import PBKDF2HMAC
    from cryptography.hazmat.primitives import hashes
except ImportError:
    print("ERROR: 'cryptography' package not installed.")
    print("Run: pip install cryptography")
    sys.exit(1)

ROOT = os.path.dirname(os.path.dirname(os.path.abspath(__file__)))
VAULT_DIR = os.path.join(ROOT, "cortex.secrets")
MANIFEST_PATH = os.path.join(VAULT_DIR, "vault.json")
LEGACY_VAULT = os.path.join(ROOT, "cortex.secrets.enc")  # DEPRECATED — do not write to this

PBKDF2_ITERATIONS = 600_000
SALT_LEN = 16
NONCE_LEN = 12


# --- Crypto primitives ---

def derive_key(passphrase: str, salt: bytes) -> bytes:
    kdf = PBKDF2HMAC(
        algorithm=hashes.SHA256(),
        length=32,
        salt=salt,
        iterations=PBKDF2_ITERATIONS,
    )
    return kdf.derive(passphrase.encode())


def encrypt_value(value: str, passphrase: str) -> bytes:
    salt = os.urandom(SALT_LEN)
    nonce = os.urandom(NONCE_LEN)
    key = derive_key(passphrase, salt)
    payload = json.dumps({
        "value": value,
        "updated_at": datetime.datetime.now(datetime.timezone.utc).isoformat()
    }).encode()
    ciphertext = AESGCM(key).encrypt(nonce, payload, None)
    return base64.b64encode(salt + nonce + ciphertext)


def _raw_decrypt(data: bytes, passphrase: str, label: str = "file") -> bytes:
    raw = base64.b64decode(data.strip())
    salt = raw[:SALT_LEN]
    nonce = raw[SALT_LEN:SALT_LEN + NONCE_LEN]
    ciphertext = raw[SALT_LEN + NONCE_LEN:]
    key = derive_key(passphrase, salt)
    try:
        return AESGCM(key).decrypt(nonce, ciphertext, None)
    except Exception:
        print(f"ERROR: Wrong passphrase or {label} is corrupt.")
        sys.exit(1)


# --- Per-file (v2) --- current format ---

def secret_path(name: str) -> str:
    safe = name.replace("/", "_").replace("\\", "_")
    return os.path.join(VAULT_DIR, f"{safe}.enc")


def ensure_vault_dir():
    os.makedirs(VAULT_DIR, exist_ok=True)


def now_iso() -> str:
    return datetime.datetime.now(datetime.timezone.utc).isoformat()


def read_manifest() -> dict:
    if os.path.exists(MANIFEST_PATH):
        with open(MANIFEST_PATH) as f:
            return json.load(f)
    return {"version": 2, "created_at": now_iso(), "passphrase_rotated_at": None, "secrets": {}}


def write_manifest(passphrase_rotated: bool = False, name: str = None, description: str = None):
    existing = read_manifest()
    # Migrate old list format to dict format
    existing_secrets = existing.get("secrets", {})
    if isinstance(existing_secrets, list):
        existing_secrets = {k: "" for k in existing_secrets}
    # Remove entries for deleted files
    names = perfile_names()
    secrets = {k: existing_secrets.get(k, "") for k in names}
    # Update description for the current key if provided
    if name and description is not None:
        secrets[name] = description
    manifest = {
        "version": 2,
        "created_at": existing.get("created_at", now_iso()),
        "passphrase_rotated_at": now_iso() if passphrase_rotated else existing.get("passphrase_rotated_at"),
        "secrets": secrets,
    }
    with open(MANIFEST_PATH, "w") as f:
        json.dump(manifest, f, indent=2)
        f.write("\n")


def perfile_names() -> list:
    if not os.path.isdir(VAULT_DIR):
        return []
    return sorted(f[:-4] for f in os.listdir(VAULT_DIR) if f.endswith(".enc") and f != "vault.json")


def perfile_get(name: str, passphrase: str) -> str:
    path = secret_path(name)
    plaintext = _raw_decrypt(open(path, "rb").read(), passphrase, label=f"cortex.secrets/{name}.enc")
    return json.loads(plaintext)["value"]


def perfile_delete(name: str, passphrase: str):
    path = secret_path(name)
    perfile_get(name, passphrase)  # verify passphrase before deleting
    os.remove(path)


# ---------------------------------------------------------------------------
# DEPRECATED: Legacy v1 single-blob support
# This code exists only to read repos that have not yet migrated from the
# original cortex.secrets.enc blob format. It will never write to the blob.
# Run `python scripts/secrets.py migrate` to eliminate this code path.
# ---------------------------------------------------------------------------

def _legacy_read(passphrase: str) -> dict:
    """DEPRECATED. Read legacy cortex.secrets.enc blob. Returns {} if not present."""
    if not os.path.exists(LEGACY_VAULT):
        return {}
    plaintext = _raw_decrypt(open(LEGACY_VAULT, "rb").read(), passphrase, label="legacy vault")
    return json.loads(plaintext)


def _legacy_delete(name: str, passphrase: str):
    """DEPRECATED. Delete a key from legacy blob by rewriting it without that key."""
    legacy = _legacy_read(passphrase)
    if name not in legacy:
        return False
    del legacy[name]
    salt = os.urandom(SALT_LEN)
    nonce = os.urandom(NONCE_LEN)
    key = derive_key(passphrase, salt)
    plaintext = json.dumps(legacy).encode()
    ciphertext = AESGCM(key).encrypt(nonce, plaintext, None)
    with open(LEGACY_VAULT, "wb") as f:
        f.write(base64.b64encode(salt + nonce + ciphertext))
    return True

# --- End deprecated legacy block ---


# --- Unified read layer (v2 first, legacy fallback) ---

def unified_list(passphrase: str = None) -> list:
    pf = set(perfile_names())
    if os.path.exists(LEGACY_VAULT):
        if passphrase is None:
            if pf:
                print("Note: legacy vault also present — provide --passphrase to include its keys.")
            legacy = {}
        else:
            legacy = _legacy_read(passphrase)  # DEPRECATED path
    else:
        legacy = {}
    all_names = pf | set(legacy.keys())
    return sorted(all_names)


def unified_get(name: str, passphrase: str) -> str:
    """Per-file first, legacy fallback (DEPRECATED path)."""
    path = secret_path(name)
    if os.path.exists(path):
        return perfile_get(name, passphrase)
    if os.path.exists(LEGACY_VAULT):  # DEPRECATED fallback
        legacy = _legacy_read(passphrase)
        if name in legacy:
            return legacy[name]
    print(f"ERROR: No secret named '{name}'.")
    sys.exit(1)


def unified_delete(name: str, passphrase: str):
    """Delete from whichever format the key lives in."""
    path = secret_path(name)
    if os.path.exists(path):
        perfile_delete(name, passphrase)
        return
    if os.path.exists(LEGACY_VAULT):  # DEPRECATED fallback
        if _legacy_delete(name, passphrase):
            return
    print(f"ERROR: No secret named '{name}'.")
    sys.exit(1)


# --- CLI commands ---

def prompt(label: str, confirm_label: str = None) -> str:
    value = getpass.getpass(f"{label}: ")
    if confirm_label:
        confirm = getpass.getpass(f"{confirm_label}: ")
        if value != confirm:
            print("ERROR: Values do not match.")
            sys.exit(1)
    return value


def cmd_store(name: str, value: str = None, passphrase: str = None, description: str = None):
    if value is None:
        value = prompt(f"Value for '{name}'")
    if not value:
        print("ERROR: Value cannot be empty.")
        sys.exit(1)
    if passphrase is None:
        existing = unified_list()
        if existing:
            passphrase = prompt("Vault passphrase")
        else:
            print("No vault found — creating a new one.")
            passphrase = prompt("Choose a passphrase", "Confirm passphrase")
    ensure_vault_dir()
    path = secret_path(name)
    blob = encrypt_value(value, passphrase)
    with open(path, "wb") as f:
        f.write(blob)
    write_manifest(name=name, description=description or "")
    print(f"Stored '{name}' -> cortex.secrets/{name}.enc")
    print("Commit and push to persist across devices.")


def cmd_get(name: str, passphrase: str = None):
    if passphrase is None:
        passphrase = prompt("Vault passphrase")
    print(unified_get(name, passphrase))


def cmd_list(passphrase: str = None):
    names = unified_list(passphrase)
    if not names:
        print("No secrets stored.")
        return
    manifest = read_manifest()
    descriptions = manifest.get("secrets", {})
    if isinstance(descriptions, list):
        descriptions = {k: "" for k in descriptions}
    print("Stored secrets:")
    for n in names:
        desc = descriptions.get(n, "")
        print(f"  {n}" + (f"  — {desc}" if desc else ""))


def cmd_delete(name: str, passphrase: str = None, force: bool = False):
    if passphrase is None:
        passphrase = prompt("Vault passphrase")
    if not force:
        confirm = input(f"WARNING: Delete '{name}'? This cannot be undone. Type the secret name to confirm: ")
        if confirm != name:
            print("Aborted.")
            sys.exit(0)
    unified_delete(name, passphrase)
    write_manifest()
    print(f"Deleted '{name}'.")


def cmd_repassphrase(old_passphrase: str = None, new_passphrase: str = None):
    names = perfile_names()
    if not names:
        print("No secrets in vault.")
        return
    if old_passphrase is None:
        old_passphrase = prompt("Current passphrase")
    # Verify old passphrase decrypts all secrets before touching anything
    values = {}
    for name in names:
        try:
            values[name] = perfile_get(name, old_passphrase)
        except SystemExit:
            print(f"ERROR: Could not decrypt '{name}' — aborting. No changes made.")
            sys.exit(1)
    if new_passphrase is None:
        new_passphrase = prompt("New passphrase", "Confirm new passphrase")
    ensure_vault_dir()
    for name, value in values.items():
        path = secret_path(name)
        blob = encrypt_value(value, new_passphrase)
        with open(path, "wb") as f:
            f.write(blob)
        print(f"  Re-encrypted '{name}'")
    write_manifest(passphrase_rotated=True)
    print(f"\nAll {len(names)} secret(s) re-encrypted with new passphrase.")
    print("Commit and push to persist across devices.")


def cmd_migrate(passphrase: str = None):
    """
    DEPRECATED path cleanup: migrate all legacy blob secrets to per-file format.
    Safe to run multiple times. Per-file secrets are not touched.
    """
    if not os.path.exists(LEGACY_VAULT):
        print("No legacy vault found. Nothing to migrate.")
        return
    if passphrase is None:
        passphrase = prompt("Vault passphrase")
    legacy = _legacy_read(passphrase)
    if not legacy:
        print("Legacy vault is empty. Nothing to migrate.")
        return
    ensure_vault_dir()
    for name, value in legacy.items():
        path = secret_path(name)
        blob = encrypt_value(value, passphrase)
        with open(path, "wb") as f:
            f.write(blob)
        print(f"  Migrated '{name}' -> cortex.secrets/{name}.enc")
    print(f"\nMigrated {len(legacy)} secret(s).")
    print("Next steps:")
    print("  1. git add cortex.secrets/")
    print("  2. git rm cortex.secrets.enc")
    print("  3. git commit -m 'vault: migrate to per-file format (v2)'")
    print("  4. git push")
    print("\ncortex.secrets.enc remains in git history (encrypted, safe).")


def main():
    parser = argparse.ArgumentParser(
        description="Cortex secrets vault (v2 per-file, v1 backwards-compatible)",
        add_help=True
    )
    sub = parser.add_subparsers(dest="cmd")

    p_store = sub.add_parser("store")
    p_store.add_argument("name")
    p_store.add_argument("--value", default=None)
    p_store.add_argument("--passphrase", default=None)
    p_store.add_argument("--description", default=None, help="Short description of what this secret is")

    p_get = sub.add_parser("get")
    p_get.add_argument("name")
    p_get.add_argument("--passphrase", default=None)

    p_list = sub.add_parser("list")
    p_list.add_argument("--passphrase", default=None)

    p_delete = sub.add_parser("delete")
    p_delete.add_argument("name")
    p_delete.add_argument("--passphrase", default=None)
    p_delete.add_argument("--force", action="store_true", help="Skip confirmation prompt")

    p_repassphrase = sub.add_parser("repassphrase", help="Re-encrypt all secrets with a new passphrase")
    p_repassphrase.add_argument("--old-passphrase", default=None)
    p_repassphrase.add_argument("--new-passphrase", default=None)

    p_migrate = sub.add_parser("migrate", help="Migrate v1 blob to per-file format (cleanup deprecated path)")
    p_migrate.add_argument("--passphrase", default=None)

    args = parser.parse_args()

    if args.cmd == "store":
        cmd_store(args.name, value=args.value, passphrase=args.passphrase, description=args.description)
    elif args.cmd == "get":
        cmd_get(args.name, passphrase=args.passphrase)
    elif args.cmd == "list":
        cmd_list(passphrase=args.passphrase)
    elif args.cmd == "delete":
        cmd_delete(args.name, passphrase=args.passphrase, force=args.force)
    elif args.cmd == "repassphrase":
        cmd_repassphrase(old_passphrase=args.old_passphrase, new_passphrase=args.new_passphrase)
    elif args.cmd == "migrate":
        cmd_migrate(passphrase=args.passphrase)
    else:
        parser.print_help()
        sys.exit(1)


if __name__ == "__main__":
    main()

File: scripts/secrets.ts
#!/usr/bin/env bun
/**
 * Cortex secrets vault — v2 per-file format with v1 legacy read support.
 *
 * Write: always per-file (cortex.secrets/<n>.enc)
 * Read:  per-file first, falls back to legacy blob (cortex.secrets.enc)
 * List:  union of both — per-file takes precedence on duplicate keys
 *
 * Binary format (must match Python implementation exactly):
 *   base64( salt[16] + nonce[12] + ciphertext_with_gcm_tag[*] )
 *   PBKDF2-SHA256, 600,000 iterations, 32-byte key
 *   AES-256-GCM — GCM tag (16 bytes) appended to ciphertext
 *
 * Run: bun scripts/secrets.ts <command> [options]
 */

import {
  pbkdf2Sync,
  createCipheriv,
  createDecipheriv,
  randomBytes,
} from 'node:crypto'
import { existsSync, readdirSync, mkdirSync, readFileSync, writeFileSync, unlinkSync } from 'node:fs'
import { join, dirname } from 'node:path'
import { fileURLToPath } from 'node:url'
import * as readline from 'node:readline'

const ROOT = join(dirname(fileURLToPath(import.meta.url)), '..')
const VAULT_DIR = join(ROOT, 'cortex.secrets')
const MANIFEST_PATH = join(VAULT_DIR, 'vault.json')
const LEGACY_VAULT = join(ROOT, 'cortex.secrets.enc')

const PBKDF2_ITERATIONS = 600_000
const SALT_LEN = 16
const NONCE_LEN = 12
const KEY_LEN = 32
const TAG_LEN = 16

// --- Crypto primitives ---

function deriveKey(passphrase: string, salt: Buffer): Buffer {
  return pbkdf2Sync(passphrase, salt, PBKDF2_ITERATIONS, KEY_LEN, 'sha256')
}

function encryptValue(value: string, passphrase: string): Buffer {
  const salt = randomBytes(SALT_LEN)
  const nonce = randomBytes(NONCE_LEN)
  const key = deriveKey(passphrase, salt)
  const payload = Buffer.from(JSON.stringify({
    value,
    updated_at: new Date().toISOString(),
  }))
  const cipher = createCipheriv('aes-256-gcm', key, nonce)
  const encrypted = Buffer.concat([cipher.update(payload), cipher.final()])
  const tag = cipher.getAuthTag()
  // Format: salt + nonce + ciphertext + tag (tag appended, matching Python AESGCM output)
  const raw = Buffer.concat([salt, nonce, encrypted, tag])
  return Buffer.from(raw.toString('base64'))
}

function rawDecrypt(data: Buffer, passphrase: string, label: string = 'file'): Buffer {
  const raw = Buffer.from(data.toString().trim(), 'base64')
  const salt = raw.subarray(0, SALT_LEN)
  const nonce = raw.subarray(SALT_LEN, SALT_LEN + NONCE_LEN)
  // ciphertext includes the GCM tag appended at the end
  const ciphertextWithTag = raw.subarray(SALT_LEN + NONCE_LEN)
  const ciphertext = ciphertextWithTag.subarray(0, ciphertextWithTag.length - TAG_LEN)
  const tag = ciphertextWithTag.subarray(ciphertextWithTag.length - TAG_LEN)
  const key = deriveKey(passphrase, salt)
  try {
    const decipher = createDecipheriv('aes-256-gcm', key, nonce)
    decipher.setAuthTag(tag)
    return Buffer.concat([decipher.update(ciphertext), decipher.final()])
  } catch {
    console.error(`ERROR: Wrong passphrase or ${label} is corrupt.`)
    process.exit(1)
  }
}

// --- Per-file (v2) ---

function secretPath(name: string): string {
  const safe = name.replace(/[/\\]/g, '_')
  return join(VAULT_DIR, `${safe}.enc`)
}

function ensureVaultDir(): void {
  mkdirSync(VAULT_DIR, { recursive: true })
}

function nowIso(): string {
  return new Date().toISOString()
}

interface Manifest {
  version: number
  created_at: string
  passphrase_rotated_at: string | null
  secrets: Record<string, string>
}

function readManifest(): Manifest {
  if (existsSync(MANIFEST_PATH)) {
    const raw = JSON.parse(readFileSync(MANIFEST_PATH, 'utf8'))
    // Migrate old list format to dict
    if (Array.isArray(raw.secrets)) {
      raw.secrets = Object.fromEntries((raw.secrets as string[]).map(k => [k, '']))
    }
    return raw as Manifest
  }
  return { version: 2, created_at: nowIso(), passphrase_rotated_at: null, secrets: {} }
}

function writeManifest(opts: { passphraseRotated?: boolean; name?: string; description?: string } = {}): void {
  const existing = readManifest()
  const names = perfileNames()
  const secrets: Record<string, string> = {}
  for (const k of names) {
    secrets[k] = existing.secrets[k] ?? ''
  }
  if (opts.name !== undefined && opts.description !== undefined) {
    secrets[opts.name] = opts.description
  }
  const manifest: Manifest = {
    version: 2,
    created_at: existing.created_at,
    passphrase_rotated_at: opts.passphraseRotated ? nowIso() : existing.passphrase_rotated_at,
    secrets,
  }
  writeFileSync(MANIFEST_PATH, JSON.stringify(manifest, null, 2) + '\n')
}

function perfileNames(): string[] {
  if (!existsSync(VAULT_DIR)) return []
  return readdirSync(VAULT_DIR)
    .filter(f => f.endsWith('.enc') && f !== 'vault.json')
    .map(f => f.slice(0, -4))
    .sort()
}

function perfileGet(name: string, passphrase: string): string {
  const path = secretPath(name)
  const data = readFileSync(path)
  const plaintext = rawDecrypt(data, passphrase, `cortex.secrets/${name}.enc`)
  return JSON.parse(plaintext.toString()).value as string
}

function perfileDelete(name: string, passphrase: string): void {
  const path = secretPath(name)
  perfileGet(name, passphrase) // verify passphrase before deleting
  unlinkSync(path)
}

// --- Legacy v1 single-blob support (DEPRECATED) ---

function legacyRead(passphrase: string): Record<string, string> {
  if (!existsSync(LEGACY_VAULT)) return {}
  const data = readFileSync(LEGACY_VAULT)
  const plaintext = rawDecrypt(data, passphrase, 'legacy vault')
  return JSON.parse(plaintext.toString()) as Record<string, string>
}

function legacyDelete(name: string, passphrase: string): boolean {
  const legacy = legacyRead(passphrase)
  if (!(name in legacy)) return false
  delete legacy[name]
  const salt = randomBytes(SALT_LEN)
  const nonce = randomBytes(NONCE_LEN)
  const key = deriveKey(passphrase, salt)
  const payload = Buffer.from(JSON.stringify(legacy))
  const cipher = createCipheriv('aes-256-gcm', key, nonce)
  const encrypted = Buffer.concat([cipher.update(payload), cipher.final()])
  const tag = cipher.getAuthTag()
  const raw = Buffer.concat([salt, nonce, encrypted, tag])
  writeFileSync(LEGACY_VAULT, Buffer.from(raw.toString('base64')))
  return true
}

// --- Unified read layer ---

function unifiedList(passphrase?: string): string[] {
  const pf = new Set(perfileNames())
  let legacy: Record<string, string> = {}
  if (existsSync(LEGACY_VAULT)) {
    if (!passphrase) {
      if (pf.size > 0) {
        console.log('Note: legacy vault also present — provide --passphrase to include its keys.')
      }
    } else {
      legacy = legacyRead(passphrase)
    }
  }
  const all = new Set([...pf, ...Object.keys(legacy)])
  return [...all].sort()
}

function unifiedGet(name: string, passphrase: string): string {
  const path = secretPath(name)
  if (existsSync(path)) return perfileGet(name, passphrase)
  if (existsSync(LEGACY_VAULT)) {
    const legacy = legacyRead(passphrase)
    if (name in legacy) return legacy[name]
  }
  console.error(`ERROR: No secret named '${name}'.`)
  process.exit(1)
}

function unifiedDelete(name: string, passphrase: string): void {
  const path = secretPath(name)
  if (existsSync(path)) {
    perfileDelete(name, passphrase)
    return
  }
  if (existsSync(LEGACY_VAULT)) {
    if (legacyDelete(name, passphrase)) return
  }
  console.error(`ERROR: No secret named '${name}'.`)
  process.exit(1)
}

// --- Interactive prompts ---

async function promptSecret(label: string, confirmLabel?: string): Promise<string> {
  if (!process.stdin.isTTY) {
    console.error('ERROR: No passphrase provided and stdin is not a TTY. Use --passphrase.')
    process.exit(1)
  }
  const rl = readline.createInterface({ input: process.stdin, output: process.stderr })
  // Hide input for passphrase-like prompts
  const hideInput = (prompt: string): Promise<string> => new Promise(resolve => {
    process.stderr.write(prompt)
    const stdin = process.openStdin()
    const tty = process.stdin as NodeJS.ReadStream & { setRawMode?: (mode: boolean) => void }
    if (tty.setRawMode) {
      let value = ''
      process.stdin.setRawMode(true)
      process.stdin.resume()
      process.stdin.setEncoding('utf8')
      const onData = (ch: string) => {
        if (ch === '\r' || ch === '\n') {
          process.stdin.setRawMode!(false)
          process.stdin.pause()
          process.stdin.removeListener('data', onData)
          process.stderr.write('\n')
          resolve(value)
        } else if (ch === '') {
          process.exit(1)
        } else if (ch === '') {
          value = value.slice(0, -1)
        } else {
          value += ch
        }
      }
      process.stdin.on('data', onData)
    } else {
      rl.question(prompt, (answer) => {
        rl.close()
        resolve(answer)
      })
    }
  })

  const value = await hideInput(`${label}: `)
  if (confirmLabel) {
    const confirm = await hideInput(`${confirmLabel}: `)
    if (value !== confirm) {
      console.error('ERROR: Values do not match.')
      process.exit(1)
    }
  }
  rl.close()
  return value
}

async function promptLine(label: string): Promise<string> {
  if (!process.stdin.isTTY) {
    console.error('ERROR: No value provided and stdin is not a TTY.')
    process.exit(1)
  }
  return new Promise(resolve => {
    const rl = readline.createInterface({ input: process.stdin, output: process.stdout })
    rl.question(`${label}: `, answer => {
      rl.close()
      resolve(answer)
    })
  })
}

// --- CLI commands ---

async function cmdStore(name: string, value?: string, passphrase?: string, description?: string): Promise<void> {
  if (!value) {
    value = await promptSecret(`Value for '${name}'`)
  }
  if (!value) {
    console.error('ERROR: Value cannot be empty.')
    process.exit(1)
  }
  if (!passphrase) {
    const existing = unifiedList()
    if (existing.length > 0) {
      passphrase = await promptSecret('Vault passphrase')
    } else {
      console.log('No vault found — creating a new one.')
      passphrase = await promptSecret('Choose a passphrase', 'Confirm passphrase')
    }
  }
  ensureVaultDir()
  const path = secretPath(name)
  const blob = encryptValue(value, passphrase)
  writeFileSync(path, blob)
  writeManifest({ name, description: description ?? '' })
  console.log(`Stored '${name}' -> cortex.secrets/${name}.enc`)
  console.log('Commit and push to persist across devices.')
}

async function cmdGet(name: string, passphrase?: string): Promise<void> {
  if (!passphrase) passphrase = await promptSecret('Vault passphrase')
  console.log(unifiedGet(name, passphrase))
}

async function cmdList(passphrase?: string): Promise<void> {
  const names = unifiedList(passphrase)
  if (names.length === 0) {
    console.log('No secrets stored.')
    return
  }
  const manifest = readManifest()
  const descriptions = manifest.secrets
  console.log('Stored secrets:')
  for (const n of names) {
    const desc = descriptions[n] ?? ''
    console.log(`  ${n}${desc ? `  — ${desc}` : ''}`)
  }
}

async function cmdDelete(name: string, passphrase?: string, force = false): Promise<void> {
  if (!passphrase) passphrase = await promptSecret('Vault passphrase')
  if (!force) {
    const confirm = await promptLine(`WARNING: Delete '${name}'? This cannot be undone. Type the secret name to confirm`)
    if (confirm !== name) {
      console.log('Aborted.')
      process.exit(0)
    }
  }
  unifiedDelete(name, passphrase)
  writeManifest()
  console.log(`Deleted '${name}'.`)
}

async function cmdRepassphrase(oldPassphrase?: string, newPassphrase?: string): Promise<void> {
  const names = perfileNames()
  if (names.length === 0) {
    console.log('No secrets in vault.')
    return
  }
  if (!oldPassphrase) oldPassphrase = await promptSecret('Current passphrase')
  const values: Record<string, string> = {}
  for (const name of names) {
    try {
      values[name] = perfileGet(name, oldPassphrase)
    } catch {
      console.error(`ERROR: Could not decrypt '${name}' — aborting. No changes made.`)
      process.exit(1)
    }
  }
  if (!newPassphrase) newPassphrase = await promptSecret('New passphrase', 'Confirm new passphrase')
  ensureVaultDir()
  for (const [name, value] of Object.entries(values)) {
    const path = secretPath(name)
    const blob = encryptValue(value, newPassphrase)
    writeFileSync(path, blob)
    console.log(`  Re-encrypted '${name}'`)
  }
  writeManifest({ passphraseRotated: true })
  console.log(`\nAll ${names.length} secret(s) re-encrypted with new passphrase.`)
  console.log('Commit and push to persist across devices.')
}

async function cmdMigrate(passphrase?: string): Promise<void> {
  if (!existsSync(LEGACY_VAULT)) {
    console.log('No legacy vault found. Nothing to migrate.')
    return
  }
  if (!passphrase) passphrase = await promptSecret('Vault passphrase')
  const legacy = legacyRead(passphrase)
  if (Object.keys(legacy).length === 0) {
    console.log('Legacy vault is empty. Nothing to migrate.')
    return
  }
  ensureVaultDir()
  for (const [name, value] of Object.entries(legacy)) {
    const path = secretPath(name)
    const blob = encryptValue(value, passphrase)
    writeFileSync(path, blob)
    console.log(`  Migrated '${name}' -> cortex.secrets/${name}.enc`)
  }
  console.log(`\nMigrated ${Object.keys(legacy).length} secret(s).`)
  console.log('Next steps:')
  console.log('  1. git add cortex.secrets/')
  console.log('  2. git rm cortex.secrets.enc')
  console.log("  3. git commit -m 'vault: migrate to per-file format (v2)'")
  console.log('  4. git push')
  console.log('\ncortex.secrets.enc remains in git history (encrypted, safe).')
}

// --- Arg parser ---

function parseArgs(argv: string[]): Record<string, string | boolean | undefined> {
  const result: Record<string, string | boolean | undefined> = {}
  for (let i = 0; i < argv.length; i++) {
    const arg = argv[i]
    if (arg.startsWith('--')) {
      const key = arg.slice(2)
      const next = argv[i + 1]
      if (next && !next.startsWith('--')) {
        result[key] = next
        i++
      } else {
        result[key] = true
      }
    } else if (!result['_cmd']) {
      result['_cmd'] = arg
    } else if (!result['_pos1']) {
      result['_pos1'] = arg
    }
  }
  return result
}

// --- Main ---

async function main(): Promise<void> {
  const argv = process.argv.slice(2)
  if (argv.length === 0) {
    console.log('Usage: bun scripts/secrets.ts <command> [options]')
    console.log('Commands: store, get, list, delete, repassphrase, migrate')
    process.exit(1)
  }

  const cmd = argv[0]
  const args = parseArgs(argv.slice(1))

  const passphrase = args['passphrase'] as string | undefined

  if (cmd === 'store') {
    const name = args['_cmd'] as string
    if (!name) { console.error('ERROR: Missing secret name.'); process.exit(1) }
    await cmdStore(name, args['value'] as string | undefined, passphrase, args['description'] as string | undefined)
  } else if (cmd === 'get') {
    const name = args['_cmd'] as string
    if (!name) { console.error('ERROR: Missing secret name.'); process.exit(1) }
    await cmdGet(name, passphrase)
  } else if (cmd === 'list') {
    await cmdList(passphrase)
  } else if (cmd === 'delete') {
    const name = args['_cmd'] as string
    if (!name) { console.error('ERROR: Missing secret name.'); process.exit(1) }
    await cmdDelete(name, passphrase, args['force'] === true)
  } else if (cmd === 'repassphrase') {
    await cmdRepassphrase(
      args['old-passphrase'] as string | undefined,
      args['new-passphrase'] as string | undefined,
    )
  } else if (cmd === 'migrate') {
    await cmdMigrate(passphrase)
  } else {
    console.error(`ERROR: Unknown command '${cmd}'.`)
    console.log('Commands: store, get, list, delete, repassphrase, migrate')
    process.exit(1)
  }
}

main().catch(err => {
  console.error(`ERROR: ${err.message}`)
  process.exit(1)
})
